rfcs/0002-mvp.md

```diff
@@ -0,0 +1,159 @@
+
+
+For the database scaffold, [this](https://gitlab.com/delibrium/delibrium-postgrest/-/tree/master/database) format is very nice, although I think we can make it a bit clearer. Here is a suggested format:
+
+```sh
+│   # Data: database fixed/seed data.
+├── data
+│   ├── some_data.{sql|csv|json}        
+│   # Models: Tables and Views
+├── models
+│   ├── public.table                    # Personal preference to explicitly add the schema, and we should prefer dot notation over nesting.
+│   │   ├── data.{sql|csv|json}         # For convenience, table specific data can be stored here too.   
+│   │   ├── model.sql                   # This is the core. We should also figure out a nice, readable format for SQL files. Should include constraints and indexes and table/column comments.
+│   │   ├── permissions.sql             # Grants and Policies.
+│   │   ├── test.sql                    # PGTAP tests?
+│   │   └── triggers.sql                # table triggers 
+│   # Functions/RPC: functions. At the moment the function are written in SQL. If we want to support something like plv8, we would have to include a builder (like webpack) to bundle the function.
+├── functions                                
+│   ├── public.func1.sql                
+│   ├── public.func2.sql│   
+│   # Postgres roles and grants.
+├── roles                                
+│   ├── role_name.sql    
+│   # Types: custom database types
+└── types
+    ├── _init.sql
+    ├── public.some_type.sql
+    └── extensions.some_type.sql
+```
+
+
+### Branching
+
+This is going to be a key part of Supabase development. Since so many of our users are on Vercel/Netlify, we should have the ability to support [branch/preview-development](https://vercel.com/docs/platform/deployments).
+
+We can do this by connecting to a different database, an idea that I got from [pgsh](https://github.com/sastraxi/pgsh). For example:
+
+- `postgresql://postgres@localhost:5432/master`
+- `postgresql://postgres@localhost:5432/develop`
+- `postgresql://postgres@localhost:5432/chore_add_types`
+
+This is the same from CI (like github actions), except that instead of running against a local database, it is running against the supabase database. We will need a way inside our config to match branch names to database names too. For example 
+
+
+### Dump
+
+```sh
+supa db dump                # dumps the currently active (local) database to disk
+supa db restore             # restores the (local) database from disk
+supa migrate                # generate the migrations file
```

Comment:
I'm still not sure how we will handle going from a declarative structure, into a set of migration scripts.

There are tools like [migra](https://github.com/djrobstep/migra) which work on a schema comparison - this might be the way to do it as long as the limitations aren't too restrictive. 

eg: "custom types/domains Basic support (drop-and-create only, no alter)". [ref](https://databaseci.com/docs/migra)

Reply:
> I'm still not sure how we will handle going from a declarative structure, into a set of migration scripts.

Ideally the developer should not look at migrations scripts and instead it should focus on declarative* SQL. So perhaps we could keep the migrations off git?(perhaps on a table)

* Of course there's a limit of how declarative SQL can be. Column renames for example always have to use ALTER.

Reply:
[metagration](https://github.com/michelp/metagration) also has an interesting approach. It allows generating up/down migration scripts with stored procedures. It also has its own table for tracking applied migrations.

Maybe we can use metagration plus migra for providing a template migration that can be edited.

Reply:
metagration also creates a restore point for down migrations, so it's possible to recover dropped columns for example(no need for a restore from a backup).

Reply:
OK cool - metagration looks interersting. 

I think it's the declarative structure -> up/down scripts that's the tricky part here though. The tools I found:

- migra (python)
- https://github.com/gimenete/dbdiff (JS, looks limited)
- https://github.com/joncrlsn/pgdiff (Go)
- https://github.com/fordfrog/apgdiff (Java)
- https://github.com/eulerto/pgquarrel (C)

Or we can of course code this directly into `postgres-meta`, perhaps using `migra` as an example since it seems the most feature complete. It will be a lot of work though.

Reply:
There's also pgadmin schema diff:

- doc: https://www.pgadmin.org/docs/pgadmin4/5.0/schema_diff.htm
- source: https://github.com/postgres/pgadmin4/tree/master/web/pgadmin/tools/schema_diff (python)

Reply:
> https://github.com/gimenete/dbdiff

Last commit on 2017.

> https://github.com/fordfrog/apgdiff

From experience(I added [RLS support](https://github.com/fordfrog/apgdiff/pull/213)), this one is not actively maintained. It has many open issues/PRs. 

> https://github.com/eulerto/pgquarrel

Has unanswered issues(also unmerged PRs), it also has partial support for many pg features: https://github.com/eulerto/pgquarrel#features. 

> https://github.com/joncrlsn/pgdiff

Actively maintained(latest commit 2 days ago). Also being Go it makes it easy to distribute.

> https://github.com/djrobstep/migra

Also has some unanswered issues, but seems the more mature of all.

---

IMO, the choice should be between pgdiff and migra.

(Maybe the pgadmin differ will become the best later)

Reply:
Seems pgadmin is already more complete than migra. Discussion here: https://github.com/djrobstep/migra/issues/148

Reply:
Some examples of how they generate their diff SQL:

- policy update: [code](https://github.com/postgres/pgadmin4/blob/master/web/pgadmin/browser/server_groups/servers/databases/schemas/tables/templates/row_security_policies/sql/default/update.sql)
- column update: [code](https://github.com/postgres/pgadmin4/blob/master/web/pgadmin/browser/server_groups/servers/databases/schemas/tables/templates/columns/sql/default/update.sql)
- trigger update: [code](https://github.com/postgres/pgadmin4/blob/master/web/pgadmin/browser/server_groups/servers/databases/schemas/tables/templates/triggers/sql/pg/default/update.sql)

Lots of work in the templates there.

Reply:
wow, there is some amazing stuff in there. Any way we can use/support this in our CLI though? I guess we would need to copy the SQL to `postgres-meta`?

Reply:
pgadmin schema diff looks pretty complete. I've noted it even does unusual cases like `ALTER VIEW v OWNER`([migra doesn't](https://github.com/djrobstep/migra/issues/160)).

![2021-03-10-164926_1351x643_scrot](https://user-images.githubusercontent.com/1829294/110702510-9a751380-81c0-11eb-9d0b-99b0fb8c1b17.png)

> Any way we can use/support this in our CLI though?

pgadmin4 seems to have a notion of CLI on their codebase, but doesn't seem to be working now. I've requested exposing the Schema Diff tool through CLI on both their mailing list and redmine - still no reply though.

If the above gets done, we could use a: `pgadmin4 --schema-diff db1 db2` in our CLI.

>  I guess we would need to copy the SQL to postgres-meta?

The SQL code I shared was the gist of the schema diff, but there's some other [logic](https://github.com/postgres/pgadmin4/blob/master/web/pgadmin/tools/schema_diff/directory_compare.py).

Translating all the logic to JS could be done, but wouldn't be better to reuse the python code? That way staying in sync with upstream changes(in case we fork it) would be easier.

Reply:
> wouldn't be better to reuse the python code?

Definitely, and nicer to support pgadmin too. Let's wait and see what they say, but very cool that the PG community are building this already.

Reply:
Unfortunately they're not open to provide a CLI mode: https://redmine.postgresql.org/issues/6304 (you'll need to sign in to see this). They mention that it would take a lot of work.

I still think the pgadmin diff is the best choice. It's already the most complete and has dedicated support. It would provide the best DX.

I've already managed to run it from source. It's a Flask application plus a Backbone.js frontend(seems they're migrating to React).

The functionality for the schema diff was started here: https://github.com/postgres/pgadmin4/commit/45f2e35a99ee691d67c45982211238b8ad40d7de. That gives me an idea of the modules involved for making it work.

How about if I try isolating that python code from the GUI and create a package of our own?

Reply:
> How about if I try isolating that python code from the GUI

Succeeded doing that. Code here: https://gist.github.com/steve-chavez/8c6825abee788394cb5424380c06487c.

Now I would need to wrap the code in a CLI.

Reply:
Nice job Steve - that's an awesome effort. 

> Now I would need to wrap the code in a CLI

Yeah, it would be very cool if we can use it here somehow.

Reply:
I have the CLI ready on this branch: https://github.com/steve-chavez/pgadmin4/blob/cli/web/cli.py.

I'm now wrapping it on a docker container so it can be tried without building the whole pgadmin4 package.

Reply:
Docker container here: https://hub.docker.com/r/stevechavez/pgadmin-schema-diff/

Usage:
```bash
docker run stevechavez/pgadmin-schema-diff \
  'postgres://postgres@host:5432/diff_source' \
  'postgres://postgres@host:5432/diff_target' \
  > diff_db.sql
```